sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala

```diff
@@ -280,6 +291,13 @@ abstract class StateStoreChangelogReader(
   }
   protected val input: DataInputStream = decompressStream(sourceStream)
 
+  private def readLineage(): Array[(Long, String)] = {
+    val lineageStr = input.readUTF()
+    Serialization.read[Array[(Long, String)]](lineageStr)
+  }
+
+  lazy val lineage: Array[(Long, String)] = readLineage()
```

Comment:
Can we represent lineage item in a case class rather than a tuple? It's hard to add a field (e.g. hasSnapshot) later with a tuple. It's also easier to read.

The way of lazily reading lineage is not robust to me. There is an implicit contract that `lineage` has to be referenced before `getNext()`. It will be better if the code doesn't build on implicit contract like this. It will be slightly better if initialization is a separate function call. Or maybe even crate a general `init()` function that does lineage reading.